File: test-suite/src/test/java/io/micronaut/kafka/docs/seek/ProductListener.java
package io.micronaut.kafka.docs.seek;

import io.micronaut.configuration.kafka.ConsumerSeekAware;
import io.micronaut.configuration.kafka.annotation.*;
import io.micronaut.configuration.kafka.seek.KafkaSeeker;
import io.micronaut.context.annotation.Requires;
import io.micronaut.kafka.docs.Product;
import org.apache.kafka.common.TopicPartition;
import java.util.*;

@KafkaListener
@Requires(property = "spec.name", value = "ConsumerSeekAwareTest")
public class ProductListener implements ConsumerSeekAware { // <1>

    List<Product> processed = new ArrayList<>();

    @Topic("awesome-products")
    void receive(Product product) {
        processed.add(product);
    }

    @Override
    public void onPartitionsRevoked(Collection<TopicPartition> partitions) { // <2>
        // save offsets here
    }

    @Override
    public void onPartitionsAssigned(Collection<TopicPartition> partitions, KafkaSeeker seeker) { // <3>
        // seek to offset here
        partitions.stream().map(tp -> seeker.seek(tp, 1)).forEach(seeker::perform);
    }
}

File: test-suite-kotlin/src/test/kotlin/io/micronaut/kafka/docs/seek/ProductListener.kt
package io.micronaut.kafka.docs.seek

import io.micronaut.configuration.kafka.ConsumerSeekAware
import io.micronaut.configuration.kafka.annotation.KafkaListener
import io.micronaut.configuration.kafka.annotation.Topic
import io.micronaut.configuration.kafka.seek.KafkaSeeker
import io.micronaut.kafka.docs.Product
import org.apache.kafka.common.TopicPartition

@KafkaListener
class ProductListener : ConsumerSeekAware { // <1>

    @Topic("awesome-products")
    fun receive(product: Product?) {
        // process product
    }

    override fun onPartitionsRevoked(partitions: Collection<TopicPartition>) { // <2>
        // save offsets here
    }

    override fun onPartitionsAssigned(partitions: Collection<TopicPartition>, seeker: KafkaSeeker) { // <3>
        // seek to offset here
        partitions.stream().map { tp -> seeker.seek(tp, 1) }.forEach(seeker::perform)
    }
}
